ip_country/src/country_block_serde.rs

```diff
@@ -341,119 +528,94 @@ impl CountryBlockDeserializer for CountryBlockDeserializerIpv6 {
-            &mut bit_queue,
-            Ipv6Addr::new(
-                0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFE,
-            ),
-        )
-        .expect("Empty BitQueue");
-        Self {
-            prev_record,
-            bit_queue,
-            empty: false,
-        }
-    }
 
-    fn get_record(bit_queue: &mut BitQueue, prev_start: Ipv6Addr) -> Option<StreamRecordIpv6> {
-        let mut segments = prev_start.segments();
-        let difference_count = (bit_queue.take_bits(3)? + 1) as usize;
-        let differences = (0..difference_count)
-            .map(|_| {
+    fn read_differences_generic(
+        bit_queue: &mut BitQueue,
+        difference_count: usize,
+        index_bit_count: usize,
+        value_bit_count: usize,
+    ) -> Vec<Difference> {
+        (0..difference_count)
+            .filter_map(|_| {
                 Some(Difference {
-                    index: bit_queue.take_bits(3)? as usize,
-                    value: bit_queue.take_bits(16)?,
+                    index: bit_queue.take_bits(index_bit_count)? as usize,
+                    value: bit_queue.take_bits(value_bit_count)?,
                 })
             })
-            .flatten()
-            .collect::<Vec<Difference>>();
-        if differences.len() < difference_count {
-            return None;
-        }
-        differences
-            .into_iter()
-            .for_each(|d| segments[d.index] = d.value as u16);
-        Some(StreamRecordIpv6 {
-            start: Ipv6Addr::from(segments),
-            country_idx: bit_queue.take_bits(9)? as usize,
-        })
+            .collect()
     }
 }
 
-#[derive(Debug)]
-struct StreamRecordIpv4 {
-    start: Ipv4Addr,
-    country_idx: usize,
-}
-
-struct StreamRecordIpv6 {
-    start: Ipv6Addr,
-    country_idx: usize,
-}
-
-struct Difference {
-    index: usize,
-    value: u64,
+struct VersionedIP<IPType, SegmentNumRep, const SEGMENTS_COUNT: usize> {
+    ip: IPType,
+    segment_num_rep: PhantomData<SegmentNumRep>,
 }
 
-fn plus_one_ipv4(ip_addr: Ipv4Addr) -> Ipv4Addr {
-    let old_data = u32_from_ipv4(ip_addr);
-    let new_data = old_data.overflowing_add(1).0;
-    Ipv4Addr::from(new_data)
+impl<IPType, SegmentNumRep, const SEGMENTS_COUNT: usize>
+    VersionedIP<IPType, SegmentNumRep, SEGMENTS_COUNT>
+{
+    fn new(ip: IPType) -> VersionedIP<IPType, SegmentNumRep, SEGMENTS_COUNT> {
+        let segment_num_rep = Default::default();
+        Self {
+            ip,
+            segment_num_rep,
+        }
+    }
 }
 
-fn minus_one_ipv4(ip_addr: Ipv4Addr) -> Ipv4Addr {
-    let old_data = u32_from_ipv4(ip_addr);
-    let new_data = old_data.overflowing_sub(1).0;
-    Ipv4Addr::from(new_data)
+impl<IPType: Debug, SegmentNumRep, const SEGMENTS_COUNT: usize> Debug
+    for VersionedIP<IPType, SegmentNumRep, SEGMENTS_COUNT>
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("VersionedIP {{ ip: {:?} }}", self.ip))
+    }
 }
```

Comment:
I just realized there would be a simpler alternative to this: 

```
#[derive(Debug)]
struct VersionedIP<IPType, SegmentNumRep, const SEGMENTS_COUNT: usize> 
where 
   IPType: Debug,
   SegmentNumRep: Debug 
{
    ip: IPType,
    segment_num_rep: PhantomData<SegmentNumRep>,
}

```

Explanation: the derive macro can be used because we added that our generics implements Debug and although I had doubted it turned out PhantomData already implements Debug. So it should work.

I really like this better, I'm sorry if it's gonna cause you troubles in your tests. You may want to delete one for your custom Debug impl.